### PR TITLE
fix(crons): Fix bugs related to seat assignment when updating a monitor

### PR DESCRIPTION
**Pull request by wedamija**
@isabellaenriquez reported that we are getting duplicate seat assignments when updating slugs, and that it's caused by us updating the slug on the monitor after we make the call to `assign_monitor_seat`.

This pr moves updating the monitor to before we update billing. While looking into this I also noticed that we will always call `assign_monitor_seat` and `disable_monitor_seat` whenever status is passed, because we don't check if the status has changed before we call them. This might not be causing any bugs but still seems incorrect, so added fixes and tests for that as well.

### CODE REVIEW

File: src/sentry/monitors/endpoints/base_monitor_details.py [hunk 1]
```diff
@@ -112,32 +112,34 @@ def update_monitor(
...
+            quotas.backend.update_monitor_slug(old_slug, params["slug"], monitor.project_id)
 
         # Attempt to assign a monitor seat
-        if params["status"] == ObjectStatus.ACTIVE:
+        if params["status"] == ObjectStatus.ACTIVE and old_status != ObjectStatus.ACTIVE:
             outcome = quotas.backend.assign_monitor_seat(monitor)
             # The MonitorValidator checks if a seat assignment is availble.
             # This protects against a race condition
             if outcome != Outcome.ACCEPTED:
                 raise ParameterValidationError("Failed to enable monitor, please try again")
```
Comment (isabellaenriquez) [inline src/sentry/monitors/endpoints/base_monitor_details.py:137]: if the update is for changing the status to active, this changes the monitor status without checking that we can actually assign a seat to that monitor, so maybe if the outcome isn't ACCEPTED we should call `monitor.update` again with the old status

File: src/sentry/monitors/endpoints/base_monitor_details.py [hunk 2]
```diff
@@ -112,32 +112,34 @@ def update_monitor(
...
+        old_slug = monitor.slug
+        old_status = monitor.status
+        if params:
+            monitor.update(**params)
+            self.create_audit_entry(
+                request=request,
+                organization=project.organization,
+                target_object=monitor.id,
+                event=audit_log.get_event_id("MONITOR_EDIT"),
+                data=monitor.get_audit_log_data(),
```
Comment (isabellaenriquez) [inline src/sentry/monitors/endpoints/base_monitor_details.py:124]: not really the purpose of this PR but for future i think it'd be nice to have more details in the audit logs since i was looking at them yesterday (iwhat is actually being updated, or at the very least the values of fields that could've been updated, ie. including status in `get_audit_log_data()`)

File: src/sentry/monitors/endpoints/base_monitor_details.py (hunk 1 above)
Comment (wedamija) [reply src/sentry/monitors/endpoints/base_monitor_details.py:137]: Hmm good point...  So ideally we should just update the slug separately ahead of time

Comment (evanpurkhiser) [reply src/sentry/monitors/endpoints/base_monitor_details.py:137]: Yeah I think we definitely need to address this.

File: src/sentry/monitors/endpoints/base_monitor_details.py [hunk 3]
```diff
@@ -112,32 +112,34 @@ def update_monitor(
         if "project" in result and result["project"].id != monitor.project_id:
             raise ParameterValidationError("existing monitors may not be moved between projects")
 
+        old_slug = monitor.slug
+        old_status = monitor.status
+        if params:
```
Comment (evanpurkhiser) [inline src/sentry/monitors/endpoints/base_monitor_details.py:117]: @wedamija if you want to pull this into it's own PR happy to approve this part ASAP

File: src/sentry/monitors/endpoints/base_monitor_details.py [hunk 4]
```diff
@@ -112,32 +112,34 @@ def update_monitor(
...
+            )
+
         # Update monitor slug
         if "slug" in result:
-            quotas.backend.update_monitor_slug(monitor.slug, params["slug"], monitor.project_id)
+            quotas.backend.update_monitor_slug(old_slug, params["slug"], monitor.project_id)
 
         # Attempt to assign a monitor seat
-        if params["status"] == ObjectStatus.ACTIVE:
+        if params["status"] == ObjectStatus.ACTIVE and old_status != ObjectStatus.ACTIVE:
```
Comment (evanpurkhiser) [inline src/sentry/monitors/endpoints/base_monitor_details.py:132]: 👍 logic seems reasonable

File: src/sentry/monitors/endpoints/base_monitor_details.py (hunk 3 above)
Comment (evanpurkhiser) [reply src/sentry/monitors/endpoints/base_monitor_details.py:131]: Misunderstood the change, ignore

**Review by evanpurkhiser — APPROVED** [review]
